hal/aurix_tc3xx.c

```diff
@@ -0,0 +1,511 @@
+
+    /* Iterate over every page in the sector, caching its contents if not
+     * erased, and caching 0s if erased */
+    for (uint32_t page = startPage; page <= endPage;
+         page += IFXFLASH_PFLASH_PAGE_LENGTH)
+    {
+        pageInSectorBuffer =
+            sectorBuffer + ((page - sectorAddress) / sizeof(uint32_t));
+
+        if (flashIsErased(page, IFXFLASH_PFLASH_PAGE_LENGTH, type)) {
+            memset(pageInSectorBuffer,
+                   FLASH_BYTE_ERASED,
+                   IFXFLASH_PFLASH_PAGE_LENGTH);
+        }
+        else {
+            readPage32Aligned(page, pageInSectorBuffer);
+        }
+    }
+}
+
+/* This function is called by the bootloader at the very beginning of the
+ * execution. Ideally, the implementation provided configures the clock settings
+ * for the target microcontroller, to ensure that it runs at at the required
+ * speed to shorten the time required for the cryptography primitives to verify
+ * the firmware images*/
+void hal_init(void)
+{
+#ifdef WOLFBOOT_AURIX_GPIO_TIMING
+    IfxPort_setPinModeOutput(&MODULE_P00,
+                             LED_WOLFBOOT,
+                             IfxPort_OutputMode_pushPull,
+                             IfxPort_OutputIdx_general);
+    IfxPort_setPinModeOutput(&MODULE_P00,
+                             LED_PROG,
+                             IfxPort_OutputMode_pushPull,
+                             IfxPort_OutputIdx_general);
+    IfxPort_setPinModeOutput(&MODULE_P00,
+                             LED_ERASE,
+                             IfxPort_OutputMode_pushPull,
+                             IfxPort_OutputIdx_general);
+    IfxPort_setPinModeOutput(&MODULE_P00,
+                             LED_READ,
+                             IfxPort_OutputMode_pushPull,
+                             IfxPort_OutputIdx_general);
+#endif /* WOLFBOOT_AURIX_GPIO_TIMING */
+
+    LED_ON(LED_WOLFBOOT);
+    LED_OFF(LED_PROG);
+    LED_OFF(LED_ERASE);
+    LED_OFF(LED_READ);
+}
+
+/*
+ * This function provides an implementation of the flash write function, using
+ * the target's IAP interface. address is the offset from the beginning of the
+ * flash area, data is the payload to be stored in the flash using the IAP
+ * interface, and len is the size of the payload. hal_flash_write should return
+ * 0 upon success, or a negative value in case of failure.
+ */
+int RAMFUNCTION hal_flash_write(uint32_t address, const uint8_t* data, int size)
+{
+    LED_ON(LED_PROG);
+
+    /* base address of containing sector (TODO what if size spans sectors?)
+     */
+    const uint32_t sectorAddress  = GET_SECTOR_ADDR(address);
+    const IfxFlash_FlashType type = getFlashTypeFromAddr(address);
+
+    /* Determine the range of pages affected */
+    const uint32_t startPage = GET_PAGE_ADDR(address);
+    const uint32_t endPage   = GET_PAGE_ADDR(address + size - 1);
+
+    /* Flag to check if sector read-modify-write is necessary */
+    bool needsSectorRmw = false;
```

Comment:
Please don't use bool. Use `int` and 1=true, 0=false.

Reply:
done